services/middleware/src/transactions.rs

```diff
@@ -0,0 +1,192 @@
+        })
+    );
+
+    checked_call.artifact_version = ">=0.9, <2".parse().unwrap();
+    assert_eq!(
+        serde_json::to_value(&checked_call).unwrap(),
+        json!({
+            "artifact_name": "test-artifact",
+            "artifact_version": ">= 0.9, < 2",
+            "inner": checked_call.inner,
+        })
+    );
+}
+
+/// Transactions executed in a batch.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[derive(ProtobufConvert, BinaryValue)]
+#[protobuf_convert(source = "proto::Batch")]
+pub struct Batch {
+    /// Transactions included in the batch.
+    pub inner: Vec<AnyTx>,
+}
+
+impl Batch {
+    /// Creates an empty batch.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Appends a call into the batch.
+    pub fn with_call(mut self, call: AnyTx) -> Self {
+        self.inner.push(call);
+        self
+    }
+}
+
+/// Transactional interface of the utilities service.
+#[exonum_interface]
+pub trait UtilsInterface {
+    /// Performs a checked call to the service. The call is dispatched only if the version
+    /// of the service matches the version requirement mentioned in the call.
+    ///
+    /// # Authorization
+    ///
+    /// The inner call is authorized in the same way as the `checked_call`.
+    fn checked_call(
+        &self,
+        context: CallContext<'_>,
+        arg: CheckedCall,
+    ) -> Result<(), ExecutionError>;
+
+    /// Performs batch execution of several transactions. Transactions are executed
+    /// in the order they are mentioned in the batch. If execution of the constituent transaction
+    /// fails, the method returns an error, thus rolling back any changes performed by earlier
+    /// transactions.
+    ///
+    /// # Authorization
+    ///
+    /// All transactions are authorized in the same way as the `batch` call itself.
+    fn batch(&self, context: CallContext<'_>, arg: Batch) -> Result<(), ExecutionError>;
+}
+
+impl UtilsInterface for MiddlewareService {
+    fn checked_call(
+        &self,
+        mut context: CallContext<'_>,
+        arg: CheckedCall,
+    ) -> Result<(), ExecutionError> {
+        let instance_id = arg.inner.call_info.instance_id;
+        let dispatcher_schema = context.data().for_dispatcher();
+        let state = dispatcher_schema
+            .get_instance(instance_id)
+            .ok_or(Error::NoService)?;
+        if state.status != InstanceStatus::Active {
+            return Err(Error::ServiceIsNotActive.into());
+        }
+
+        let artifact = &state.spec.artifact;
+        if arg.artifact_name != artifact.name {
+            return Err(Error::ArtifactMismatch.into());
+        }
+        if !arg.artifact_version.matches(&artifact.version) {
+            return Err(Error::VersionMismatch.into());
+        }
+
+        context
+            .call_context(instance_id, ChildAuthorization::Fallthrough)?
+            .call("", arg.inner.call_info.method_id, arg.inner.arguments)
```

Comment:
Should we place a TODO here to update this place once interface names will become a part of the `AnyTx`?